Avoid treating null RPC arguments different from missing arguments

ryanofsky opened the pull request:
This is a followup to #10783.

- The first commit doesn't change behavior at all, just simplifies code.
- The second commit just changes RPC methods to treat null arguments the same as missing arguments instead of throwing type errors.
- The third commit updates developer notes after the cleanup.
- The forth commit does some additional code cleanup in `getbalance`.

Followup changes that should happen in future PRs:

- [ ] Replace uses of `.isTrue()` with calls to `.get_bool()` so numbers, objects, and strings cause type errors instead of being interpreted as false. https://github.com/bitcoin/bitcoin/pull/11050#discussion_r133850525
- [ ] Add braces around if statements. https://github.com/bitcoin/bitcoin/pull/11050#discussion_r133851133
- [ ] Maybe improve UniValue type error exceptions and eliminate RPCTypeCheck and RPCTypeCheckArgument functions. https://github.com/bitcoin/bitcoin/pull/11050#discussion_r133829303

ryanofsky committed:
Get rid of redundant RPC params.size() checks

No change in behavior.

ryanofsky committed:
Avoid treating null RPC arguments different from missing arguments

This changes RPC methods to treat null arguments the same as missing arguments,
instead of throwing type errors. Specifically:

- `getbalance` method now returns the wallet balance when the `account` param
  is null instead of throwing a type error (same as when parameter is missing).
  It is still an error to supply `minconf` or `watchonly` options when the
  account is null.

- `addnode` and `setban` methods now return help text instead of type errors if
  `command` params are null (same as when params are missing).

- `sendrawtransaction`, `setaccount`, `movecmd`, `sendfrom`,
  `addmultisigaddress`, `listaccounts`, `lockunspent` methods accept null
  default values where missing values were previously allowed, and treat them
  the same.

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2361,19 +2361,18 @@ UniValue lockunspent(const JSONRPCRequest& request)
         return true;
     }
 
+    RPCTypeCheckArgument(request.params[1], UniValue::VARR);
```
`get_array` already throws and above you also replace `RPCTypeCheck` validation with `get_bool`. Remove?

ryanofsky replied:
> get_array already throws and above you also replace RPCTypeCheck validation with get_bool. Remove?

It throws a different error that doesn't specify passed type name.

promag replied:
So add `RPCTypeCheckArgument` for 1st param?

promag replied:
Sorry, you did!

jnewbery replied:
> It throws a different error that doesn't specify passed type name.

So we should update `is_<type>()` error message to output the passed type.

ryanofsky replied:
> So we should update is_<type>() error message to output the passed type.

If you mean `get_<type>()` error message, we can do this but it would involve making a change upstream in https://github.com/jgarzik/univalue. Anyway this commit is trying to do exactly one thing: "treat null arguments the same as missing arguments, instead of throwing type errors". If people don't like RPCTypeCheck functions and would prefer to rely on UniValue exceptions, it really seems like a followup for a separate PR.

jnewbery replied:
yes, I meant `get_<type>()`.

Agree that this should be done in a follow-up PR. I've tested that this PR maintains existing behaviour for the error checking in `lockunspent`

ryanofsky replied:
Added as followup in https://github.com/bitcoin/bitcoin/pull/11050#issue-250186439

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2361,19 +2361,18 @@ UniValue lockunspent(const JSONRPCRequest& request)
     bool fUnlock = request.params[0].get_bool();
 
-    if (request.params.size() == 1) {
+    if (request.params[1].isNull()) {
```
IMO `size()` sounds better in this case.

ryanofsky replied:
The problem with checking size == 1 is that it's fragile and will break if you want to add new arguments to the RPC. 

Checking params.size() can also create confusion when using named arguments, because the current named argument implementation creates variable length null-padded arrays, so adding or removing one named argument can cause a completely different argument to switch from missing to null or vice versa.

The point of this PR is to avoid uses of params.size() to prevent missing arguments from being treated differently than null arguments.

promag replied:
If another argument is added then this would be `if (request.params[1].isNull() && request.params[2].isNull())` right?

ryanofsky replied:
Thread https://github.com/bitcoin/bitcoin/pull/11050#discussion_r133097936

No, params[1] is the `transactions` argument to the `lockunspent` RPC. The `transactions` argument may be an array of transaction outputs, or it may be null, or it may be missing.

This code is just saying that if the `transactions` argument is null or missing, then all transactions should be unlocked. The previous code only worked if `transactions` argument was missing, but not if the `transactions` argument was null (because it would trigger a type error).

laanwj replied:
Agree with ryanofsky. `isNull` is more correct here.

promag reviewed (commented):
utACK 4195613 modulo nits.

ryanofsky committed:
Update developer notes after params.size() cleanup

ryanofsky commented:
Added a documentation commit 4195613ea411c164e8b30c5803d35ea90a863dee -> f4c29d6e9d11885f522ea05499261ba0da56af36 ([pr/narg.2](https://github.com/ryanofsky/bitcoin/commits/pr/narg.2) -> [pr/narg.3](https://github.com/ryanofsky/bitcoin/commits/pr/narg.3), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/narg.2...pr/narg.3))

promag commented:
utACK f4c29d6.

ryanofsky committed:
Clean up getbalance RPC parameter handling

Only change in behavior is that unsupported combinations of parameters now
trigger more specific error messages instead of the vague "JSON value is not a
string as expected" error.

ryanofsky commented:
Added another commit to clean up getbalance code a little and make it return nicer errors:

f4c29d6e9d11885f522ea05499261ba0da56af36 -> a53cb75c9f8a8aca54830039bc1ed7f160a6e30c ([pr/narg.3](https://github.com/ryanofsky/bitcoin/commits/pr/narg.3) -> [pr/narg.4](https://github.com/ryanofsky/bitcoin/commits/pr/narg.4), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/narg.3...pr/narg.4)).

MeshCollider commented:
utACK https://github.com/bitcoin/bitcoin/commit/745d2e315f39d7591c0ea9e772a19e3cd9b51b09

TheBlueMatt commented on src/rpc/rawtransaction.cpp:
```diff
@@ -922,7 +922,7 @@ UniValue sendrawtransaction(const JSONRPCRequest& request)
 
     CAmount nMaxRawTxFee = maxTxFee;
-    if (request.params.size() > 1 && request.params[1].get_bool())
+    if (!request.params[1].isNull() && request.params[1].get_bool())
```
Couldnt this just be request.params[1].isTrue() like many other checks?

jnewbery replied:
This is safer. It throws an error when the user passes in "True" as a string (rather than silently evaluating to false)

laanwj replied:
> rather than silently evaluating to false

Yes, please be careful to throw exceptions on wrong types instead of silently ignoring errors. Than can lead to hazardous bugs in user code.

TheBlueMatt commented on src/wallet/rpcwallet.cpp:
```diff
@@ -838,11 +851,11 @@ UniValue movecmd(const JSONRPCRequest& request)
     if (nAmount <= 0)
         throw JSONRPCError(RPC_TYPE_ERROR, "Invalid amount for send");
-    if (request.params.size() > 3)
+    if (!request.params[3].isNull())
```
How about we just stop type-checking it instead?

ryanofsky replied:
> How about we just stop type-checking it instead?

I think there's some benefit to keeping the type checking here, and it doesn't really cost anything. Maybe somebody forgets the dummy parameter and is calling this RPC with a comment. Slightly better to throw an error in get_int instead of silently dropping the comment.

jnewbery reviewed (approved):
utACK 745d2e315f39d7591c0ea9e772a19e3cd9b51b09

I especially like the changes to `getbalance`. Returning an error is preferable to silently ignoring arguments passed in.

ryanofsky reviewed (commented):
Thanks for the review.

> I especially like the changes to getbalance. Returning an error is preferable to silently ignoring arguments passed in.

Note that the arguments weren't silently ignored before. They would just trigger more opaque "JSON value is not a string as expected" errors if they were non-null. (This is mentioned in the commit message.)

jnewbery reviewed (approved):
ACK

promag commented on src/rpc/rawtransaction.cpp:
```diff
@@ -339,14 +339,14 @@ UniValue createrawtransaction(const JSONRPCRequest& request)
     }
 
-    bool rbfOptIn = request.params.size() > 3 ? request.params[3].isTrue() : false;
+    bool rbfOptIn = request.params[3].isTrue();
```
This param should be type checked?

ryanofsky replied:
> This param should be type checked?

Added as followup in https://github.com/bitcoin/bitcoin/pull/11050#issue-250186439. This commit is not changing behavior in any way, and the PR is not adding new type checking in other places.

promag replied:
> This commit is not changing behavior

Right!

promag commented on src/rpc/net.cpp:
```diff
@@ -193,7 +193,7 @@ UniValue getpeerinfo(const JSONRPCRequest& request)
 {
     std::string strCommand;
-    if (request.params.size() == 2)
+    if (!request.params[1].isNull())
```
Add `{}` where appropriate.

ryanofsky replied:
> Add {} where appropriate.

Added as followup in https://github.com/bitcoin/bitcoin/pull/11050#issue-250186439. This PR is already doing a bunch of different things and has been reviewed by 4 people. I don't want to add noise, delay, and complexity sticking braces everywhere as well.

promag replied:
Sure.

promag commented:
utACK 745d2e3.

TheBlueMatt commented:
utACK 745d2e315f39d7591c0ea9e772a19e3cd9b51b09

laanwj commented:
utACK 745d2e3
